R/input-time-series.R

```diff
@@ -565,3 +567,167 @@ build_hierarchy_label <- function(meta_areas) {
+        difference = value_spectrum - value_naomi) |>
+    dplyr::select(indicator, area_name, year, group,
+                  value_spectrum, value_naomi, difference)
+}
+
+##' Compare aggregated subnational ART inputs + spectrum totals for comparison table
+##'
+##' @param art Path to file containing ART data or ART data object
+##' @param shape Path to file containing geojson areas data or areas data object
+##' @param pjnz Path to zip file containing spectrum pjnz file/s
+##' @keywords internal
+prepare_anc_spectrum_comparison <- function(anc, shape, pjnz) {
+
+  ## Check if shape is object or file path
+  if(!inherits(shape, "sf")) {
+    shape <- read_area_merged(shape) }
+
+  ## Check if anc is object or file path
+  if(!inherits(anc, c("spec_tbl_df","tbl_df","tbl","data.frame" ))) {
+    anc <- read_anc_testing(anc)
+  }
+
+  ## PJNZ either object or file path
+  if (!inherits(pjnz, "spec_program_data")) {
+    pjnz <- extract_pjnz_program_data(pjnz) }
+
+  ## Aggregate ART data
+  anc_agreggated <- anc |>
+    dplyr::left_join(shape,  by = "area_id") |>
+    tidyr::pivot_longer(dplyr::starts_with("anc"),
+                        names_to = "indicator",
+                        values_to = "value_naomi") |>
+    dplyr::count(spectrum_region_code, age_group, year, indicator,
+                 wt = value_naomi, name = "value_naomi")
+
+  ## Aggregate Spectrum data
+  spec_aggregated <- pjnz$anc_testing |>
+        dplyr::rename("value_spectrum" = "value")
+
+  # Get spectrum level to select correct area names
+  spectrum_region_code <- unique(shape$spectrum_region_code)
+
+  if(length(spectrum_region_code) > 1){spectrum_level <- 1}else{spectrum_level <- 0}
+
+  dat  <- dplyr::left_join(anc_agreggated, spec_aggregated,
+                           by = c("spectrum_region_code", "year", "indicator")) |>
+    dplyr::left_join(shape |>
+                       dplyr::filter(area_level == spectrum_level) |>
+                       dplyr::select(area_name, spectrum_region_code),
+                     by = "spectrum_region_code") |>
+    dplyr::filter(indicator %in% unique(pjnz$anc_testing$indicator))
+
+  #  Return data formatted for comparison table
+  dat |>
+    dplyr::mutate(
+      sex = "female", age_group = "Y015_049",
+      group = "anc_adult_female",
+      difference = value_spectrum - value_naomi) |>
+    dplyr::select(indicator, area_name, year, group,
+                  value_spectrum, value_naomi, difference)
+
+}
+
+##' Compare aggregated subnational Naomi + spectrum totals for comparison table
+##'
+##' @param art Path to file containing ART data or ART data object
+##' @param anc Path to file containing ART data or ART data object
+##' @param shape Path to file containing geojson areas data or areas data object
+##' @param pjnz Path to zip file containing spectrum pjnz file/s
+##' @keywords internal
+
+prepare_spectrum_naomi_comparison <- function(art, anc, shape, pjnz){
```

Comment:
I think we need to handle the case when either anc or art is missing? Users could upload 1 but not the other.

I'd change signature to
```
prepare_spectrum_naomi_comparison(shape, pjnz, art = NULL, anc = NULL)
```

Then have a check,

```
if (!is.null(art)) {
  if(!inherits(art, c("spec_tbl_df","tbl_df","tbl","data.frame" ))) {
    art <- read_art_number(art, all_columns = TRUE)
  }
  art_comparison <- prepare_art_spectrum_comparison(art, shape, pjnz)
}
```

and similar for ART. Then only need to bind if both present.

Reply:
Can you add a quick test for this too please? You can mock our `prepare_art_spectrum_comparison` and `prepare_anc_spectrum_comparison` to just test that it reads the data and that handles case above if ANC or ART are missing

Reply:
Have added both and to produce an empt table when user provides no programme data. Kept `rbind` in to be able to bind empty df.